### ISSUE TITLE
$meteor.object sends entire object on update of single property

### ISSUE THREAD

**Issue by cnelson711** [Urigo/angular-meteor#349]
I'd like to have something like:
```
<input ng-model="currentUser.profile.name" type="text">
```
But because a $meteor.object $sets the entire object on change, the request is denied because of security restrictions on the non-profile properties of the user.

```
Lines 125-132
angular-meteor-object.js
============
        }, function (newItem, oldItem) {
          if (newItem !== oldItem && newItem) {
            var newItemId = newItem._id;
            if (newItemId && !_.isEmpty(newItem = _.omit(angular.copy(newItem), '_id'))) {

              // Instead of updating the entire object, why not do diffObject and
              // only $set those properties that changed
              collection.update({_id: newItemId}, {$set: newItem});
            }
          }
        }, true);
```

I can work around the issue with something like this:
```
index.ng.html
==============
<input ng-model="currentUser.profile.name" type="text" ng-keyup="loginCtrl.syncUsername()">


controller.ng.js
=================
ctrl.syncUsername = function() {
    Meteor.users.update({_id:Meteor.user()._id}, {$set {"profile.name":$rootScope.currentUser.profile.name}});
};
```

**Comment by Urigo** [comment]
Yes that's true.
It's something we have on $meteor.collection.

I starting to see that there is no real difference between $meteor.collection and $meteor.object as people use nested collections in both.

$meteor.object started just to simplify the user's code a bit, because you can achieve the same thing with $meteor.collection very easily.
I'm thinking of merging the code of $meteor.object to just use $meteor.collection 

Would love thoughts and help on that

**Comment by auterium** [comment]
@Urigo I do use $meteor.object in the resolve part of states to determine if the page should be displayed or to show a 404 page instead. Here's an example of how I use it:
```javascript
.state('my.state', {
  url: '/state/:id',
  resolve: {
    statesSub: ['$meteor', '$stateParams', function($meteor, $stateParams) {
      return $meteor.subscribe('states-single', $stateParams.id);
    },
    state: ['$meteor', '$stateParams', '$q', 'statesSub', function($meteor, $stateParams, $q) {
      var obj = $meteor.object(States, $stateParams.id, false);
      if(!obj._id) {
        obj.stop();
        return $q.reject('NOT_FOUND');
      }
      return obj;
    }]
  }
});
```
And then use an error catcher to display a 404 page:
```javascript
.run(['$rootScope', '$state', '$meteor', function($rootScope, $state, $meteor) {
  $rootScope.$on('$stateChangeError', function(event, toState, toParams, fromState, fromParams, error) {
    switch(error) {
      case 'AUTH_REQUIRED':
        return $state.go('login');
      case 'FORBIDDEN':
        return $state.go('403-page');
      case 'NOT_FOUND':
        return $state.go('404-page');
    }
  });
}]);
```

What if $meteor.oject is kept but that it only updates the required fields instead of the full doc?

**Comment by Urigo** [comment]
@auterium I would love your help in thinking how to refactor $meteor.object with the following goals:

1. make it based on $meteor.collection to avoid code duplicates
2. Move it's functionality functions like reset and getRawObject to $meteor.collection so they both will enjoy it
3. Keep the $meteor.object functionality as is (with the improvements of $meteor.collection like nested copy)

After all, find one is just:
https://github.com/meteor/meteor/blob/devel/packages/minimongo/minimongo.js#L128-L144

**Comment by Urigo** [comment]
Also, for a short term solution, we can just copy the logic of diffArray from the collection to object.
@ fredrikekelund created that great nested solution, what do you think?

**Comment by Urigo** [comment]
Also, for a short term solution, we can just copy the logic of diffArray from the collection to object.
@fredrikekelund created that great nested solution, what do you think?

**Comment by auterium** [comment]
Hi @Urigo, I could do that, just this week I'm very busy as this weekend I'll got to a startup weekend, so I have to get ahead on my regular job. After this weekend I'll be able to work on it having $meteor.object to user $meteor.collection.

**Comment by fredrikekelund** [comment]
Been a while since I looked closer at the code, but based on what I remember and a quick glance at the source I think reusing diffArray shouldn't be too hard, maybe we can just expose any internal methods that we need to make a diff for `$meteor.object` specifically

**Comment by Urigo** [comment]
@auterium have you had further thoughts on that?

**Comment by auterium** [comment]
Hey @Urigo, here are some thoughts:
- Make all items within $meteor.collection to be instances of SubObject
- Rename all SubObject's methods to start with $ to avoid property collitions
- $meteor.object would internally call $meteor.collection and return the first item only

Let me know what you think and if we agree I'll do the code.

**Comment by Urigo** [comment]
@auterium I'm not sure exactly about the first 2 points, can you share an example?

**Comment by auterium** [comment]
Ok, so right now when you use $meteor.object, you get an instance of SubObject, which has the getRawObject, save, reset, etc methods, so basically $meteor.collection would do the same for each object in the array.

For the second point, though is not that common, but let's say that for some reson you want to have a boolean property called reset, which common sense would say to use allowReset or isReset, but I'm too lazy to do that, so I just want to call it reset. There I'll have a name collision with the property and the method, while if the method to reset the object would be called $reset, then there would be no collision as mongo doesn't allow property names to start with $. Another use case would be that I need to store JSON objects that come from an external source (i.e. 3rd party website) where I have no control of the property names and if any of those properties are named save, reset, etc, I'll have to create a converter or always be using getRawObject and in both cases completely forget about the autosave option.

Hope I made myself clear and didn't confused you more lol

**Comment by Urigo** [comment]
@auterium yes that's very smart.
My worry is about performance on the collection.  do you think it will affect?
Maybe we can think of a way to make it an object only on demand?

**Comment by auterium** [comment]
I had my worries about performance too, so I was thinking that that could be an option when calling $meteor.collection which by default would be false. Here's a proposed signature:
$meteor.collection(collection, [autoClientSave, [updateCollection, [returnSubObject]]])

I'm not very fan of adding lots of arguments to methods, so this is something I would like more:
$meteor.collection(collection, [autoClientSave|options])
Where if the second arg is an object, then it's treated as an options object, if it's a boolean then it's for autoClientSave, and otherwise it throws an error.

A third option would be to have sort of 2 signatures, where if the second argument is an object, then it's used as an options object and otherwise it keeps workin as it is right now, to avoid breaking existing functionality.

**Comment by Urigo** [comment]
@auterium I like the options parameter, that's really nice and also follows Meteor's conversions (kind of like Mongo..)
My thinking was (not sure) to maybe at the beginning just to return the object with the functions just when calling $meteorObject (without changing collection itself)
but I'm not sure if it's possible with the implementation and just saw that as a first step.

What do you think?

**Comment by auterium** [comment]
@Urigo, sorry for the late response, I've been very busy lately. I checked the source code and found that adding SubObject support for $meteorCollection is not a straight-forward thing to do, so I think that for now the way to go will be to make $meteorObject depend on $meteorCollection and use it.

**Comment by Urigo** [comment]
ok sounds good.
thank you so much for your efforts and time!

**Comment by Urigo** [comment]
@cnelson711 can you test again against the latest changes in master?

**Comment by DAB0mB** [comment]
See [this](https://github.com/DAB0mB/angular-meteor/commit/75fc5543133c8114d312e22ed03736711b4905bd) commit, soon to be merged

**Comment by Urigo** [comment]
Fixed in 1.0.0-rc.5
